### PR TITLE
[madSHaLz] APOC triggers aren't updated after a user deletes a database

### PR DESCRIPTION
**Pull request by vga91**
Implemented 3rd solution - `DatabaseEventListener`,
so that we can delete every system node belonging to the database that we delete.

Added a `cleanUpSystemNodes()` for backward compatibility

### CODE REVIEW

File: core/src/main/java/apoc/cypher/CypherInitializer.java [hunk 1]
```diff
@@ -70,6 +88,15 @@ public void available() {
                     } catch (Exception ignored) {
                         userLog.info("Cannot check APOC version compatibility because of a transient error. Retrying your request at a later time may succeed");
                     }
+
+                    // this block is already in system db, so we can directly execute system operation
+                    // for backward compatibility: clear system nodes at start-up if the corresponding database name doesn't exist
+                    // placed here instead of e.g. `ApocConfig` because at this point we are sure that all databases are recovered
+                    cleanUpSystemNodes();
```
Comment (ncordon) [inline core/src/main/java/apoc/cypher/CypherInitializer.java:95]: I don't think this is true. Databases are brought up in an eventual consistent way in 5.x so when system is up neo4j could not be up

File: it/src/test/java/apoc/it/core/TriggerEnterpriseFeaturesTest.java [hunk 2]
```diff
@@ -46,10 +49,15 @@ public class TriggerEnterpriseFeaturesTest {
...
+        final String cypherInitializer = String.format("%s.%s.0",
+                APOC_CONFIG_INITIALIZER, SYSTEM_DATABASE_NAME);
+        final String createInitDb = String.format("CREATE DATABASE %s IF NOT EXISTS", INIT_DB);
+
         // We build the project, the artifact will be placed into ./build/libs
         neo4jContainer = createEnterpriseDB(List.of(TestContainerUtil.ApocPackage.CORE), true)
                 .withEnv(APOC_TRIGGER_ENABLED, "true")
-                .withEnv(TRIGGER_REFRESH, String.valueOf(TRIGGER_DEFAULT_REFRESH));
+                .withEnv(TRIGGER_REFRESH, String.valueOf(TRIGGER_DEFAULT_REFRESH))
+                .withEnv(cypherInitializer, createInitDb);
```
Comment (ncordon) [inline it/src/test/java/apoc/it/core/TriggerEnterpriseFeaturesTest.java:60]: Why do we need this?

File: core/src/test/java/apoc/trigger/TriggerRestartTest.java [hunk 3]
```diff
@@ -62,6 +74,49 @@ private void restartDb() {
...
+        restartDb();
+
+        withSystemDb(tx -> {
+            Iterator<Node> dbNotExistentNodes = tx.findNodes(ApocTrigger, database.name(), dbName);
+            assertFalse( dbNotExistentNodes.hasNext() );
+
+            Iterator<Node> neo4jNodes = tx.findNodes(ApocTrigger, database.name(), db.databaseName());
+            assertTrue( neo4jNodes.hasNext() );
+        });
+    }
```
Comment (ncordon) [inline core/src/test/java/apoc/trigger/TriggerRestartTest.java:111]: I think we should add a test where we install a trigger for neo4j, drop neo4j and eventually the trigger nodes are wiped out

Comment (vga91) [reply core/src/test/java/apoc/trigger/TriggerRestartTest.java:111]: I added it in `TriggerEnterpriseFeaturesTest.testDeleteTriggersAfterDatabaseDeletion`

File: it/src/test/java/apoc/it/core/TriggerEnterpriseFeaturesTest.java (hunk 2 above)
Comment (vga91) [reply it/src/test/java/apoc/it/core/TriggerEnterpriseFeaturesTest.java:60]: Sorry, I had put some extra code that I probably mistakenly deleted after a rebase :X
It's to check that even if I create a db via `apoc.initializer.*` and then delete it, the implementation still works.
See the test `TriggerEnterpriseFeaturesTest.testDeleteTriggersAfterDatabaseDeletion` (line 193)

File: core/src/main/java/apoc/cypher/CypherInitializer.java (hunk 1 above)
Comment (vga91) [reply core/src/main/java/apoc/cypher/CypherInitializer.java]: Yes indeed it is not safe.
I've been looking for other ways to do this but I don't think there is currently a method that isn't completely error prone, but maybe I'm missing something.

Perhaps it's better not to do this thing at all and document somewhere that for backward compatibility we have to delete the triggers manually. Wdyt?

File: core/src/test/java/apoc/trigger/TriggerRestartTest.java [hunk 4]
```diff
@@ -10,19 +13,28 @@
...
 import static apoc.ApocConfig.SUN_JAVA_COMMAND;
+import static apoc.SystemLabels.ApocTrigger;
+import static apoc.SystemPropertyKeys.database;
+import static apoc.SystemPropertyKeys.name;
 import static apoc.trigger.TriggerTestUtil.TRIGGER_DEFAULT_REFRESH;
 import static apoc.trigger.TriggerTestUtil.awaitTriggerDiscovered;
 import static apoc.util.TestUtil.waitDbsAvailable;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
```
Comment (ncordon) [inline core/src/test/java/apoc/trigger/TriggerRestartTest.java:37]: I think these imports are not needed anymore?

File: it/src/test/java/apoc/it/core/TriggerEnterpriseFeaturesTest.java [hunk 5]
```diff
@@ -150,7 +158,57 @@ public void testTriggerInstallInNewDatabase() {
...
+            );
+
+            // drop database
+            sysSession.writeTransaction(tx -> tx.run(String.format("DROP DATABASE %s WAIT;", INIT_DB)));
+
+            // check that the trigger has been removed
+            testCallEmpty(sysSession, "CALL apoc.trigger.show($dbName)",
+                    Map.of("dbName", INIT_DB)
+            );
+        }
```
Comment (ncordon) [inline it/src/test/java/apoc/it/core/TriggerEnterpriseFeaturesTest.java:210]: Can you split this into another test please?

Comment (vga91) [reply it/src/test/java/apoc/it/core/TriggerEnterpriseFeaturesTest.java]: Separated

**Review by ncordon — APPROVED** [review]

**Comment by ncordon** [comment]
Feel free to merge, we can fix the Snyk issues in another PR, as they are related to dependencies that they've just deemed are insecure

**Pull request merged**